### PR TITLE
Improve Playwright integration

### PR DESCRIPTION
**Pull request by baev**
1. add message & trace to the steps with errors
2. add repetition parameter to test results
3. move history id calculation to the end of the test and include all
   the parameters (except included) to it.
4. rework testing. Use fork instead of spawn and message results to the
   parent (instead of using stdio)
5. configure allure reporter for the tests
6. add additional metadata (steps, attachments) to the test results

#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

File: packages/allure-playwright/test/fixtures.ts [hunk 1]
```diff
@@ -76,53 +81,81 @@ const runPlaywrightTest = async (
...
     },
     cwd: baseDir,
   });
-  let output = "";
-  testProcess.stdout.on("data", (chunk) => {
-    output += String(chunk);
+  const results: AllureResults = { tests: [], groups: [], attachments: {} };
+  testProcess.on("message", (message) => {
+    const event: { path: string; type: string; data: string } = JSON.parse(message.toString());
+    // console.log(event);
```
Comment (vovsemenv) [inline packages/allure-playwright/test/fixtures.ts:101]: delete

File: packages/allure-playwright/src/index.ts [hunk 2]
```diff
@@ -168,16 +180,9 @@ class AllureReporter implements Reporter {
...
-      let trace = result.error.stack && stripAscii(result.error.stack);
-      if (trace && message && trace.startsWith(`Error: ${message}`)) {
-        trace = trace.substr(message.length + "Error: ".length);
-      }
-      allureTest.statusDetails = {
-        message,
-        trace,
-      };
+    const error = result.error;
+    if (error) {
```
Comment (vovsemenv) [inline packages/allure-playwright/src/index.ts:184]: maybe its topic for totally different PR but we also have `results.error` which is array of results and we need to process all the errors in result

File: packages/allure-playwright/src/index.ts [hunk 3]
```diff
@@ -307,15 +310,12 @@ class AllureReporter implements Reporter {
...
-        const writer = this.allureWriter;
-        void writer; // Used in `eval()`, below.
-        const postProcess = eval(process.env.PW_ALLURE_POST_PROCESSOR_FOR_TEST); // eslint-disable-line no-eval
-        console.log(JSON.stringify(postProcess(this.allureWriter))); // eslint-disable-line no-console
-      } catch (e) {
-        console.log(JSON.stringify({ error: (e as Error).stack || String(e) })); // eslint-disable-line no-console
-      }
+
+    if (this.options.environmentInfo) {
+      this.allureRuntime?.writeEnvironmentInfo(this.options?.environmentInfo);
```
Comment (vovsemenv) [inline packages/allure-playwright/src/index.ts:315]: why `onEnd`?

File: packages/allure-playwright/src/index.ts [hunk 4]
```diff
@@ -57,7 +65,7 @@ class AllureReporter implements Reporter {
   options: AllureReporterOptions;
 
   private allureWriter = process.env.PW_ALLURE_POST_PROCESSOR_FOR_TEST
```
[the remark this answers is not in the document]
Comment (baev) [reply packages/allure-playwright/src/index.ts:67]: we should override the writer in generated config instead. Plan to deal with that in separate PR

File: packages/allure-playwright/src/index.ts (hunk 2 above)
Comment (baev) [reply packages/allure-playwright/src/index.ts:184]: different PR, yes

File: packages/allure-playwright/src/index.ts (hunk 3 above)
Comment (baev) [reply packages/allure-playwright/src/index.ts:315]: All the misc files should be generated at the end of the run to allow changing them in the tests/fixtures. Currently, there is no runtime API for that, but I moved it just to align with the vision that such data should be delivered at the end

File: packages/allure-playwright/test/fixtures.ts (hunk 1 above)
Comment (baev) [inline packages/allure-playwright/test/fixtures.ts:101]:
```suggestion
```

File: packages/allure-playwright/test/stdio.spec.ts [hunk 5]
```diff
@@ -36,24 +35,66 @@ test("should report stdout and stderr", async ({ runInlineTest }) => {
...
-    },
-    {
-      name: "stderr",
-      type: "text/plain",
-      buffer: "System err 1\nSystem err 2\n{ foo: 'bar' }\n[ 1, 2, 3, 'test' ]\nSystem err 3\n",
-    },
+});
+
+test("should report stderr", async ({ runInlineTest }) => {
+  test.skip(true, "stderr is not reported by playwright for some reason");
```
Comment (epszaw) [inline packages/allure-playwright/test/stdio.spec.ts:61]: What's wrong with the test? Should we keep it skipped?

File: packages/allure-playwright/test/fixtures.ts (hunk 1 above)
Comment (epszaw) [inline packages/allure-playwright/test/fixtures.ts]: Unnecessary comment

**Review by epszaw — APPROVED** [review]
✅ Tested locally. Works as before. @vovsemenv could you please check the PR's revision with your playwright examples?
Left some comments. Good to fix/discuss, but not necessary.
Great job! 👍

**Review by vovsemenv — APPROVED** [review]
looks fine, works on my example pw tests
image

**Review by vovsemenv — APPROVED** [review]
looks fine, works on my example pw tests
image

File: packages/allure-js-commons/src/writers/MessageAllureWriter.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    sendData(distFileName, "attachment", readFileSync(from));
+  }
+
+  writeEnvironmentInfo(info?: Record<string, string | undefined>): void {
+    const text = stringify(info, { unicode: true }).toString();
+    sendData("environment.properties", "misc", Buffer.from(text));
+  }
+
+  writeCategoriesDefinitions(categories: Category[]): void {
+    writeJson("categories.json", "misc", categories);
```
[the remark this answers is not in the document]
Comment (baev) [reply packages/allure-js-commons/src/writers/MessageAllureWriter.ts:48]: yep

File: packages/allure-js-commons/src/AllureTest.ts [hunk 7]
```diff
@@ -49,4 +50,23 @@ export class AllureTest extends ExecutableItemWrapper {
   addTmsLink(url: string, name: string): void {
     this.addLink(url, name, LinkType.TMS);
   }
+
+  calculateHistoryId(): void {
+    const tcId = this.testResult.testCaseId
```
Comment (mxschmitt) [inline packages/allure-js-commons/src/AllureTest.ts:55]: fyi: we have a history ID already exposed: https://playwright.dev/docs/api/class-testcase#test-case-id

**Pull request merged**